### PR TITLE
Shapes example

### PR DESCRIPTION
**Pull request by scrouthtv**
TLDR: How to create a window and draw / fill things using `xproto`:
image
(the white border outside the black border is generated by my WM, should not be there for you)

A bit of backstory: I originally got into `xgb` because I was working on a go implementation of the `xwd` (x window dump) file format and I needed to create small windows with randomly colored pixels, something like this:
image
So I got into `xproto` and how to create images and such.

I cloned this example originally from [Alex Kesling's testing](https://github.com/akesling/Gonads/blob/master/sandbox/xgb_shapes.go) which is basically a Go copy of an example in the [`xcb` documentation](https://xcb.freedesktop.org/tutorial/basicwindowsanddrawing/).

I adapted some things to go nicer along the `create-window` example, added commentary and some more things (e. g. drawing text and filling a rectangle).

### CODE REVIEW

**Comment by scrouthtv** [comment]
When I first was trying to use `xproto`, I couldn't figure out how I'd get the drawable that was needed for drawing.

I wanted one of the core aspects of this example to be demonstrating that the drawable is (for very basic use cases) the window itself.

File: examples/shapes/main.go [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+	// that should be used to draw something. All available properties
+	//
+	// These properties can be set by or'ing their keys (xproto.Gc*)
+	// and adding the value to the end of the values array.
+	// The order in which the values have to be given corresponds to the order that they defined
+	// mentioned in `xproto`.
+	//
+	// Here we create a new graphics context
+	// which only has the foreground (color) value set to black:
+	foreground, _ := xproto.NewGcontextId(X)
```
Comment (jezek) [inline examples/shapes/main.go:57]: Please check for error. I know, there is an error only if all IDs are depleted, and is impossible to happen in this example, but we don't want be a bad example for go learners. Always handle errors.

File: examples/shapes/main.go [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+	mask := uint32(xproto.GcForeground)
+	values := []uint32{screen.BlackPixel}
+	xproto.CreateGC(X, foreground, draw, mask, values)
+
+	// It is possible to set the foreground value to something different.
+	// In production, this should use xorg color maps instead for compatibility
+	// but for demonstration setting the color directly also works.
+	// For more information on color maps, see the xcb documentation:
+	// https://x.org/releases/X11R7.5/doc/libxcb/tutorial/#usecolor
+	red, _ := xproto.NewGcontextId(X)
```
Comment (jezek) [inline examples/shapes/main.go:67]: Handle error, please.

File: examples/shapes/main.go [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+	mask = uint32(xproto.GcForeground)
+	values = []uint32{0xff0000}
+	xproto.CreateGC(X, red, draw, mask, values)
+
+	// We'll create another graphics context that draws thick lines:
+	thick, _ := xproto.NewGcontextId(X)
```
Comment (jezek) [inline examples/shapes/main.go:73]: Handle error, please.

**Review by jezek — CHANGES_REQUESTED** [review]
I took the libery to fix typos and reword a comment.
I had some difficulties running the example on Ubuntu (see comments near font).

Note: As you can see, I was able to push commit directly to your master branch. This is because you gave me permission to edit your PR. It's better practice to start a PR from a branch, so the maintainer (if granted privileges) can only edit the branch (and not master directly).

**Comment by scrouthtv** [comment]
What are we going to do about the missing font issue? 

The example font that is often used is simply `fixed` (which is afaik installed by X).

However, I also want to demonstrate writing UTF8 characters (the ö), but `fixed` does only have a very limited character set.

I tried finding a font that would be preinstalled on many systems, looks like that didn't work out.

It is possible to see the available fonts using `xfontsel`.
`xproto` can also list available fonts to the developer using `ListFonts()`.

However, I don't know how to determine which of these support unicode characters.

**Comment by scrouthtv** [comment]
Could you kindly try the font spec
`-*-*-*-*-*-*-14-*-*-*-*-*-iso10646-1`

**Comment by jezek** [comment]
> However, I also want to demonstrate writing UTF8 characters (the ö), but `fixed` does only have a very limited character set.

I don't have a problem with `fixed`.
shapes fixed

But with unicode and fonts it's a little bit tricky (I think). If I wanted "more unicode" demonstration, I would pick a text like ```Hellö 世界!```, where the `fixed` font will fail like the `-*-*-*-*-*-*-14-*-*-*-*-*-iso10646-1` font (and maybe the `-gnu-unifont-*-*-*-*-16-*-*-*-*-*-*-*` too).
shapes iso10646 jp

I'm a little bit in the dark, how to handle this, there is no easy solution (If there is, I would like to know). Few thing come into my mind, like bundle font with app and/or use enviroment variables, input lags or some config file to let the user choose desired font. But for simplicity of this example, I would go with `fixed` and  leave a comment about some possible drawbacks.

File: examples/shapes/main.go [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+		{X: 90, Y: 100, Width: 55, Height: 40, Angle1: 20 << 6, Angle2: 270 << 6},
+	}
+
+	for {
+		evt, err := X.WaitForEvent()
+
+		if err != nil {
+			fmt.Println("error reading event:", err)
+			return
+		} else if evt == nil && err == nil {
```
Comment (jezek) [inline examples/shapes/main.go:205]: Here ```} else if evt == nil {``` is enough. No need to test for `err == nil`, cause we are in else clause of `err != nil`.

**Comment by scrouthtv** [comment]
> I don't have a problem with `fixed`.

I just tried again, and now it's working for me too. I guess I just had something wrong during my first tries.

> `-*-*-*-*-*-*-14-*-*-*-*-*-iso10646-1`

This simply selectes any font that is iso10646-1 encoded, which is basically the character encoding of Unicode. So any of these fonts should have at least a basic set of extended latin characters.

> But for simplicity of this example, I would go with fixed and leave a comment about some possible drawbacks.

Sounds good!
